gordon_janitor_gcp/cloud_dns.py

```diff
@@ -0,0 +1,125 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2018 Spotify AB
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Client module to interact with the Google Cloud DNS API.
+
+This client makes use of the asynchronus HTTP client as defined in
+:py:mod:`gordon_janitor_gcp.http_client`, and therefore must use
+service account/JWT authentication (for now).
+
+To use:
+
+.. code-block:: pycon
+
+    >>> keyfile = '/path/to/service_account_keyfile.json'
+    >>> client = AIOGoogleDNSClient(
+    ...   project='my-dns-project', keyfile=keyfile)
+    >>> records = client.get_records_for_zone('testzone')
+    >>> print(records[0])
+    GCPResourceRecordSet(name='foo.testzone.com', type='A',
+                         rrdatas=['10.1.2.3'], ttl=300)
+
+"""
+
+import logging
+
+import attr
+
+from gordon_janitor_gcp import http_client
+
+
+@attr.s
+class GCPResourceRecordSet:
+    """DNS Resource Record Set.
+
+    Args:
+        name (str): Name/label.
+        type (str): Record type (see `Google's supported records
+            <https://cloud.google.com/dns/overview#supported_dns_r
+            ecord_types>`_ for valid types).
+        rrdatas (list): Record data according to RFC 1034§3.6.1 and
+            RFC 1035§5.
+        ttl (int): (optional) Number of seconds that the record set can
+            be cached by resolvers. Defaults to 300.
+    """
+    # TODO (lynn): This will be moved to a common package to be shared
+    #   between all of gordon* packages. It will also make use of attrs
+    #   ability to optionally validate upon creation.
+    name     = attr.ib(type=str)
+    type     = attr.ib(type=str)
+    rrdatas  = attr.ib(type=list)
+    ttl      = attr.ib(type=int, default=300)
+
+
+class AIOGoogleDNSClient(http_client.AIOGoogleHTTPClient):
+    """Async HTTP client to interact with Google Cloud DNS API.
+
+    Attributes:
+        BASE_URL (str): base call url for the DNS API
+
+    Args:
+        project (str): Google project ID that hosts the managed DNS.
+        keyfile (str): path to service account (SA) keyfile.
+        scopes (list): scopes with which to authorize the SA. Default is
+            ``['cloud-platform']``.
+        api_version (str): DNS API endpoint version. Defaults to ``v1``.
+        loop: asyncio event loop to use for HTTP requests.
+    """
+    BASE_URL = 'https://www.googleapis.com/dns'
```

Comment:
(idea!) What do you think about creating a class variable for scopes? I don't think the client will be instantiated with different scopes than the ones it needs, I think, so they could be part of the class definition. Then, `__init__` could work with this by doing something like `super().__init__(keyfile=keyfile, scopes=self.SCOPES, loop=loop)`.

Reply:
There are three potential scopes to use with DNS. read, read/write, and a general cloud-platform (which we default to if nothing, and seems like is applicable to all cloud APIs). I'd still like to give the option to choose which to use. Thoughts?

Reply:
Sounds good to me, I didn't think that the DNS API is a different beast from the APIs I'm working with, which only ever need one scope each to do their job.